nni/algorithms/nas/pytorch/fbnet/trainer.py

```diff
@@ -0,0 +1,417 @@
+            # init mutator
+            self.mutator = FBNetMutator(model, lookup_table)
+
+        # DataParallel should be put behind the init of mutator
+        self.model = torch.nn.DataParallel(self.model, device_ids=device_ids)
+        self.model.to(device)
+
+        if self.arch_search:
+            # build architecture optimizer
+            self.arch_optimizer = torch.optim.AdamW(
+                self.mutator.get_architecture_parameters(),
+                self.config.nas_lr,
+                weight_decay=self.config.nas_weight_decay,
+            )
+            self.reg_loss = RegularizerLoss(config=self.config)
+
+        self.criterion = criterion
+        self.epoch = 0
+
+    def _layer_choice_sample(self):
+        """
+        sample the index of network within layer choice
+        """
+        stages = [stage_name for stage_name in self.lookup_table.layer_num]
+        stage_lnum = [self.lookup_table.layer_num[stage] for stage in stages]
+
+        # get the choice idx in each layer
+        choice_ids = list()
+        layer_id = 0
+        for param in self.mutator.get_architecture_parameters():
+            param_np = param.cpu().detach().numpy()
+            op_idx = np.argmax(param_np)
+            choice_ids.append(op_idx)
+            self.logger.info(
+                "layer {}: {}, index: {}".format(layer_id, param_np, op_idx)
+            )
+            layer_id += 1
+
+        # get the arch_sample
+        choice_names = list()
+        layer_id = 0
+        for i, stage_name in enumerate(stages):
+            ops_names = [op for op in self.lookup_table.lut_ops[stage_name]]
+            for j in range(stage_lnum[i]):
+                searched_op = ops_names[choice_ids[layer_id]]
+                choice_names.append(searched_op)
+                layer_id += 1
+
+        self.logger.info(choice_names)
+        return choice_names
+
+    def _validate(self):
+        """
+        Do validation. During validation, LayerChoices use the mixed-op.
+
+        Returns
+        -------
+        float, float, float
+            average loss, average top1 accuracy, average top5 accuracy
+        """
+        self.valid_loader.batch_sampler.drop_last = False
+        batch_time = AverageMeter("batch_time")
+        losses = AverageMeter("losses")
+        top1 = AverageMeter("top1")
+        top5 = AverageMeter("top5")
+
+        # test on validation set under eval mode
+        self.model.eval()
+
+        end = time.time()
+        with torch.no_grad():
+            for i, (images, labels) in enumerate(self.valid_loader):
+                images = images.to(self.device, non_blocking=True)
+                labels = labels.to(self.device, non_blocking=True)
+
+                if self.arch_search:
+                    perf_cost = Variable(torch.zeros(self.dev_num, 1)).to(
+                        self.device, non_blocking=True
+                    )
+                    output, _ = self.model(images, self.temp, perf_cost)
```

Comment:
So the supernet takes 3 parameters: `images`, `temperature` and `perf_cost`? Looks a burden to user and such a supernet is not compatible with other trainers.

Reply:
The gumbel softmax will be calculated with temperature; and the total flops of current chosen sub-net will also be returned as the perf_cost.

Reply:
So what happens if I use a supernet from DARTS example in your trainer?

It's not runnable at all because the forward signature doesn't match. Right?

Reply:
Done, the temperature is updated by calling mutator.set_temperature(); and the accumulation of perf_cost is moved outside the forward function.